indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -56,33 +59,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
 
   public Optional<SupervisorSpec> getSupervisorSpec(String id)
   {
-    return supervisors.get(id) == null
-           ? Optional.<SupervisorSpec>absent()
-           : Optional.fromNullable(supervisors.get(id).rhs);
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorSpec>absent() : Optional.fromNullable(supervisor.rhs);
   }
 
-  public boolean hasSupervisor(String id)
-  {
-    return supervisors.containsKey(id);
-  }
-
-  public boolean createAndStartSupervisor(SupervisorSpec spec)
+  public boolean createOrUpdateAndStartSupervisor(SupervisorSpec spec)
   {
+    Preconditions.checkState(started, "SupervisorManager not started");
```

Comment:
@drcrallen if it's inside the lock, won't you have that problem you were talking about where the lifecycle start blocks and Jetty exhausts its threads because this method can't obtain the lock? The reason it's outside of the lock is to address that comment.